dbld based release automation

bazsi opened the pull request:
This branch adds support for both release preparation and the actual release process into dbld as well as making git an optional dependency for dbld in a number of cases.

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build SUCCESS

bazsi commented:
This is how you can test it:

```
$ ./dbld/rules prepare-release VERSION=3.29.1
### the command above should prepare the source tree for the release of 3.29.1 and leave it as local changes

$ ./dbld/rules release VERSION=3.29.1
### this command should validate that the source tree is well prepared for 3.29.1 and there are no local changes
### once this checks out, it generates a release with 3.29.1 as version number, builds rpm/deb and tags the release.
```

Both options should be documented in the help screen.

bazsi commented:
rebased against latest master.

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build SUCCESS

mitzkia commented:
Executing the following command `./dbld/rules prepare-release VERSION=3.29.1`
add change to the following file also, but it should not: `tests/python_functional/functional_tests/parsers/app-parser/test_topic_syslog_raw.py`

the change is:
```diff
-    (r"""<159>Dec 19 10:48:57 EST 10.203.28.21 vendor=Websense product=Security product_version=7.7.0 action=permitted severity=1 category=153 user=- src_host=10.64.134.74 src_port=62189 dst_host=mail.google.com dst_ip=74.125.224.53 dst_port=443 bytes_out=197 bytes_in=76 http_response=200 http_method=CONNECT http_content_type=- http_user_agent=Mozilla/5.0_(Windows;_U;_Windows_NT_6.1;_enUS;_rv:1.9.2.23)_Gecko/20110920_Firefox/3.6.23 http_proxy_status_code=200 reason=- disposition=1034 policy=- role=8 duration=0 url=https://mail.google.com""", "${.app.name}", "websense"),
+    (r"""<159>Dec 19 10:48:57 EST 10.203.29.21 vendor=Websense product=Security product_version=7.7.0 action=permitted severity=1 category=153 user=- src_host=10.64.134.74 src_port=62189 dst_host=mail.google.com dst_ip=74.125.224.53 dst_port=443 bytes_out=197 bytes_in=76 http_response=200 http_method=CONNECT http_content_type=- http_user_agent=Mozilla/5.0_(Windows;_U;_Windows_NT_6.1;_enUS;_rv:1.9.2.23)_Gecko/20110920_Firefox/3.6.23 http_proxy_status_code=200 reason=- disposition=1034 policy=- role=8 duration=0 url=https://mail.google.com""", "${.app.name}", "websense"),
```

mitzkia commented:
I would add a successful execution check to the new commands: prepare-release and release
Maybe after each sub-command like: update_version, update_packaging_debian, update_packaging_rhel, update_version_refs_in_source.

From my experience: 
- I have not got a build specific environment (without installed any specific build packages)
- When I executed the following command: `./dbld/rules prepare-release VERSION=3.29.1`, it has finished almost immediately
- For the first glance I have not recognized that the output contains an error:
```console
$ ./dbld/rules prepare-release VERSION=3.29.1
/home/micek/git_projects/github_mitzkia/syslog-ng/dbld/prepare-release 3.29.1
Updating VERSION file
Updating Debian packaging
/home/micek/git_projects/github_mitzkia/syslog-ng/dbld/prepare-release: line 20: debchange: command not found
Updating RHEL packaging
Updating version references in source files
The proposed version changes have been saved as local changes. Review and commit them or abort using 'git reset --hard'.
```

Because this tool has a high risk in the release process, I would suggest to add internal checks between the executed commands if they are returns successfully (with 0 return code).

mitzkia commented:
I am not sure this changes are also valid, executed command: `./dbld/rules prepare-release VERSION=3.29.1`
- git status output after executing the command
```console
        ....
	typechange: packaging/debian/syslog-ng-mod-mongodb.postinst
	typechange: packaging/debian/syslog-ng-mod-mongodb.postrm
	typechange: packaging/debian/syslog-ng-mod-mongodb.preinst
	typechange: packaging/debian/syslog-ng-mod-sql.postinst
	typechange: packaging/debian/syslog-ng-mod-sql.postrm
	typechange: packaging/debian/syslog-ng-mod-sql.preinst
       ....
```

Before executing the command these files was previously symlinks, but after the change they are regular files.

mitzkia commented:
For me `./dbld/rules release VERSION=3.29.1` command works.

```console
...
+ cd /home/micek/rpmbuild/BUILD
+ cd syslog-ng-3.29.1
+ /usr/bin/rm -rf /home/micek/rpmbuild/BUILDROOT/syslog-ng-3.29.1-1.el7.x86_64
+ exit 0
Your rpms are in /build/centos-7, also available on the host in $(top_srcdir)/dbld/build/centos-7
Building the release was successful, artifacts stored in /home/micek/git_projects/github_mitzkia/syslog-ng/dbld/release/3.29.1
/usr/bin/git tag syslog-ng-3.29.1
```

bazsi commented:
I have now pushed another iteration where:

* I have fixed error handling in prepare-release (I kind of expected the "set -e" to kick in, but it seems not to be inherited into functions).
* I have changed rules so that prepare-release is running inside the TARBALL image, so the necessary tools are there (with one extra addition to dbld/packages.manifest)

kira-syslogng commented:
Build SUCCESS

mitzkia commented:
With the latest changes I have the following experiences.
1, `./dbld/rules prepare-release VERSION=3.29.1` prints the following warnings from `update_packaging_debian` function (prepare-release currently runs from balabit/syslog-ng-ubuntu-bionic docker image), maybe it should be fixed:
```
libdistro-info-perl is not installed, Debian release names are not known.
libdistro-info-perl is not installed, Ubuntu release names are not known.
debchange warning: Unable to determine if unstable is a valid Ubuntu release.
```
2, Unnecessary version changes for: `test_topic_syslog_raw.py` and for `scl/websense/plugin.conf` files are gone -> it is ok
3, Unnecessary file type changes for: `*.postinst`, `*.postrm`, `*.preinst` are gone -> it is ok

bazsi commented:
On Fri, Jul 10, 2020, 14:14 mitzkia <notifications@github.com> wrote:

> With the latest changes I have the following experiences.
> 1, ./dbld/rules prepare-release VERSION=3.29.1 prints the following
> warnings from update_packaging_debian function (prepare-release currently
> runs from balabit/syslog-ng-ubuntu-bionic docker image), maybe it should be
> fixed:
>
> libdistro-info-perl is not installed, Debian release names are not known.
> libdistro-info-perl is not installed, Ubuntu release names are not known.
> debchange warning: Unable to determine if unstable is a valid Ubuntu release.
>
> If you rebuild your image this should be fixed. This package was added to
packages.manifest

>  2, Unnecessary version changes for: test_topic_syslog_raw.py and for
> scl/websense/plugin.conf files are gone -> it is ok
> 3, Unnecessary file type changes for: *.postinst, *.postrm, *.preinst are
> gone -> it is ok
>
I think your first comment should also be resolved and with that all
outstanding issues were addressed.


—
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/syslog-ng/syslog-ng/pull/3293#issuecomment-656644273>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFOK5WBWTZXEYSHM24UPI3R24A23ANCNFSM4NC5WOVA>
> .
>

bazsi commented:
@lbudai can you please perform a review?

lbudai commented on dbld/prepare-release:
```diff
@@ -0,0 +1,64 @@
+	echo Updating version references in source files
+	ESCAPED_OLD_MAJOR=$(echo $OLD_MAJOR | sed -e 's/[]\/$*.^[]/\\&/g') &&
+	ESCAPED_NEW_MAJOR=$(echo $NEW_MAJOR | sed -e 's/[\/&]/\\&/g') &&
+	git ls-files $VERSION_FILE_LIST | xargs sed -i "s/$ESCAPED_OLD_MAJOR/$ESCAPED_NEW_MAJOR/g"
```
@bazsi : I opened #3368 - for addressing this note -  If that PR merged, then I'll modify the unit tests and test configs.
At the end we won't need that huge sed, we can reduce its scope.

bazsi replied:
Please note that the tool didn't commit the patch unconditionally, rather it just presented it to the user performing the release. However, it is true that in the long term it would be great if prepare-release could become completely automated.

The "@version: current" idea is a good one, however I don't thing we should document it. User configuration should never use it as we would be losing the upgrade mechanism.

bazsi replied:
Also, please note that the latest version of the patch didn't change lib/tests/test_lexer.c, the list of files to be changed are listed in this variable:

VERSION_FILE_LIST="README.md doc/man/*.xml
        scl/syslog-ng.conf
        contrib/openbsd-packaging/syslog-ng.conf
        packaging/debian/syslog-ng.conf
        packaging/rhel/syslog-ng.conf"

alltilla replied:
I thought of an alternative of the `sed` in cases, where `version: current` cannot be used.

We could create a `.in` file for every file, the config version is used. The `.in` file would have some kind of templating, like `@@CONFIG_VERSION@@`, which would be resolved in its `Makefile.am`. It would be enough to just create a list, which gathers files like these, and add each file to that list. The list would then be processed in the main `Makefile.am`. In the end it would be a `sed`, yes, but it would be much cleaner. Of course the same should be done in cmake.

If you agree with this, I am happy to implement it.

kira-syslogng commented:
Build SUCCESS

bazsi commented:
Hi,

Back in a day, we used to have what you suggest, that's how packaging files
were maintained for instance.

I think sometimes preprocessing at configure time makes things a bit more
confusing (by for instance the actual file not available in git)

This is the list of files that contain the version number:

VERSION_FILE_LIST="README.md doc/man/*.xml
scl/syslog-ng.conf
contrib/openbsd-packaging/syslog-ng.conf
packaging/debian/syslog-ng.conf
packaging/rhel/syslog-ng.conf"

Most of these are config files used by various packaging scripts. The rest
is README.md and the manual pages.

I think this should be good enough, unless the others have other opinions.
For them I just want to remind them of the version number logic we used to
have in autoconf.

On Thu, Aug 6, 2020, 08:56 Attila Szakacs <notifications@github.com> wrote:

> *@alltilla* commented on this pull request.
> ------------------------------
>
> In dbld/prepare-release
> <https://github.com/syslog-ng/syslog-ng/pull/3293#discussion_r466185059>:
>
> > +
> +function update_packaging_debian() {
> +	echo Updating Debian packaging
> +	DEB_FULLNAME=$USER_NAME DEB_EMAIL=$USER_EMAIL debchange --distribution unstable --newversion "$VERSION-1" -c packaging/debian/changelog "New upstream release $VERSION"
> +}
> +
> +function update_packaging_rhel() {
> +	echo Updating RHEL packaging
> +	sed -i -e "/^%changelog/ a * `date '+%a %b %e %Y'` $USER_NAME <$USER_EMAIL> - $VERSION-1\n- updated to $VERSION\n" packaging/rhel/syslog-ng.spec
> +}
> +
> +function update_version_refs_in_source() {
> +	echo Updating version references in source files
> +	ESCAPED_OLD_MAJOR=$(echo $OLD_MAJOR | sed -e 's/[]\/$*.^[]/\\&/g') &&
> +	ESCAPED_NEW_MAJOR=$(echo $NEW_MAJOR | sed -e 's/[\/&]/\\&/g') &&
> +	git ls-files $VERSION_FILE_LIST | xargs sed -i "s/$ESCAPED_OLD_MAJOR/$ESCAPED_NEW_MAJOR/g"
>
> I thought of an alternative of the sed in cases, where version: current
> cannot be used.
>
> We could create a .in file for every file, the config version is used.
> The .in file would have some kind of templating, like @@CONFIG_VERSION@@,
> which would be resolved in its Makefile.am. It would be enough to just
> create a list, which gathers files like these, and add each file to that
> list. The list would then be processed in the main Makefile.am. In the
> end it would be a sed, yes, but it would be much cleaner. Of course the
> same should be done in cmake.
>
> If you agree with this, I am happy to implement it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/syslog-ng/syslog-ng/pull/3293#discussion_r466185059>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFOK5WSC7CS3ZQCGSKWZRTR7JH2DANCNFSM4NC5WOVA>
> .
>

alltilla commented:
@bazsi:
We have other README files, which have the version burnt in:
```
[15:52][ ~/Work/repos/OSE ] (master) $ git grep '3\.28' | grep README
README.md:@version: 3.28
README.md:@version: 3.28
README.md:@version: 3.28
modules/http/README.md:@version: 3.28
modules/java-modules/common/README.md:@version: 3.28
modules/java/README.md:@version: 3.28
modules/java/README.md:@version: 3.28
scl/graphite/README:@version: 3.28
scl/syslogconf/README:@version: 3.28
```
I think these should be updated as well with every version bump.

I think when a new `README` file is added to the `scl`s, no one will think about adding that to the `dbld` script, and its version will be stuck. For this reason, I don't think the `sed` in the `dbld` script is future-proof, which an automated script should aim to be.

I would add the generated files, too to `git`, the only confusing thing would be to make sure to only edit the template files, and not the generated ones.

bazsi commented:
Do we need those module specific README.md files? Sometimes they may come handy while encountering a module the first time I try to test/hack things. But since we don't install these anywhere, I don't think leaving those version numbers behind (or even removing them completely) would be a real problem.

I think most of those README files originate from times where that specific plugin was external (in incubator for instance). Apart from the one in the root directory, 

I don't think those README files should be maintained automatically. (this is in contrast with manual pages or default configs which are user-visible and should be maintained).

kira-syslogng commented:
Build SUCCESS
